Fix resizable panel ratio issue when resizing

FadhlanR opened the pull request:
Ticket: CS-6168

Previously, we encountered difficulty maintaining the ratio of resizable panels during resizing. This was due to the absence of a recorded ratio to guide the adjustment of panel sizes when `didResize` is triggered. In this PR, I've introduced a new `panelRatio` field. With this addition, we can now leverage the `panelRatio` field to determine panel sizes accurately during resizing operations.

Before:

https://github.com/cardstack/boxel/assets/12637010/968fd842-6fd5-432e-97c2-b2860eb5b246


After:

https://github.com/cardstack/boxel/assets/12637010/b494e518-1e6e-401c-9f77-9fe02b8d12ec

FadhlanR committed:
Fix resizable panel ratio issue when resizing

backspace commented on packages/boxel-ui/addon/src/components/resizable-panel-group/index.gts:
```diff
@@ -155,11 +176,29 @@ export default class ResizablePanelGroup extends Component<Signature> {
     return id;
   }
 
+  @action
```
I think this doesn’t need to be `@action` since it looks like it’s only called from internal functions vs the template

backspace commented:
I’ll check this out more fully later today but it’s looking good, from the video it seems like at the most-compressed the bottom panel is still bigger than the top one, any idea why that is?

FadhlanR commented:
> I’ll check this out more fully later today but it’s looking good, from the video it seems like at the most-compressed the bottom panel is still bigger than the top one, any idea why that is?

Thanks in advance @backspace  !, the bottom panel is bigger because it has a minimum length.

backspace commented:
> the bottom panel is bigger because it has a minimum length.

ah yes, understood. Maybe that should be removed or updated somehow, it’s meant to have a default height that shows ≈3.5 files, this is difficult to calculate in a general way because of varying viewport heights though. But the top panel is supposed to be the priority.

Does it make sense to calculate the ratio at rendering time (or if a handle has been dragged, resetting it) and then applying it when the container changes? Here it looks like the bottom panel is just shrinking vs the ratio being preserved:

![screencast 2023-11-15 11-08-26](https://github.com/cardstack/boxel/assets/43280/a6551edb-31c7-4e47-ae1d-c87deb9409b9)

There are so many variables here 😯

FadhlanR commented:
> > the bottom panel is bigger because it has a minimum length.
> 
> ah yes, understood. Maybe that should be removed or updated somehow, it’s meant to have a default height that shows ≈3.5 files, this is difficult to calculate in a general way because of varying viewport heights though. But the top panel is supposed to be the priority.
> 
> Does it make sense to calculate the ratio at rendering time (or if a handle has been dragged, resetting it) and then applying it when the container changes? Here it looks like the bottom panel is just shrinking vs the ratio being preserved:
> 
> ![screencast 2023-11-15 11-08-26](https://user-images.githubusercontent.com/43280/283177407-a6551edb-31c7-4e47-ae1d-c87deb9409b9.gif) [ ![screencast 2023-11-15 11-08-26](https://user-images.githubusercontent.com/43280/283177407-a6551edb-31c7-4e47-ae1d-c87deb9409b9.gif) ](https://user-images.githubusercontent.com/43280/283177407-a6551edb-31c7-4e47-ae1d-c87deb9409b9.gif) [ ](https://user-images.githubusercontent.com/43280/283177407-a6551edb-31c7-4e47-ae1d-c87deb9409b9.gif)
> 
> There are so many variables here 😯

So, the top panel must always be larger than the bottom panel, correct? If that's the case, I propose removing the minimum length constraint on the bottom panel. When you mention the default size of the bottom panel showing '3.5 files', does this size need to be maintained as the container gets smaller, or is it only necessary the first time a user opens the code submode?

I don't believe we can accurately calculate the ratio during resizing. After `didResize` is triggered, the panel sizes are affected by the resizing, and it's possible that one of the panels becomes too small compared to the others due to their minimum lengths. If we use these sizes to calculate the ratio and then resize the container back to its normal size, one of the panels may end up being too small.

lukemelia commented on packages/boxel-ui/addon/src/components/resizable-panel-group/index.gts:
```diff
@@ -365,109 +408,50 @@ export default class ResizablePanelGroup extends Component<Signature> {
-      return;
+    let remainingContainerSize = newContainerSize - sumArray(panelLengths);
+    let panelWithNoMinLength = Array.from(this.listPanelContext).find(
+      (panelContext) => panelContext[1].minLengthPx == undefined,
```
panelContext name is confusing because this is a tuple with an integer and a PanelContext, not a PanelContext

lukemelia commented on packages/boxel-ui/addon/src/components/resizable-panel-group/index.gts:
```diff
@@ -365,109 +408,50 @@ export default class ResizablePanelGroup extends Component<Signature> {
+      panelLengths[0] = panelLengths[0]! + remainingContainerSize;
+    } else if (remainingContainerSize < 0 && panelWithNoMinLength) {
+      panelLengths[panelWithNoMinLength[0] - 1] =
+        panelLengths[panelWithNoMinLength[0] - 1]! + remainingContainerSize > 0
```
consider extracting a local variable or two to simplify this

lukemelia commented on packages/boxel-ui/addon/src/components/resizable-panel-group/index.gts:
```diff
@@ -365,109 +408,50 @@ export default class ResizablePanelGroup extends Component<Signature> {
-        this.listPanelContext.set(largestPanel.id, {
-          ...largestPanelContext,
-          lengthPx: largestPanel.length + panelGroupWithoutResizeHandlerLength,
+    for (let index = 1; index <= this.listPanelContext.size; index++) {
```
consider either a) using a zero based position system, or b) naming this "position" instead of "index"

lukemelia replied:
lean toward zero-based for everything -- it's what developers expect

backspace commented:
> the default size of the bottom panel showing '3.5 files', does this size need to be maintained as the container gets smaller, or is it only necessary the first time a user opens the code submode

It was a comment during a meeting and we didn’t fully go into it but I think it was just a guideline for a “reasonable” window size, not something that needed to be maintained.

But yes it does seem like if there’s a minimum size at all it should be on the top panel.

FadhlanR committed:
Update resizing logic

FadhlanR committed:
Remove unnecessary action

FadhlanR committed:
lint fixes

FadhlanR commented:
@lukemelia @backspace I've implemented several updates as follows:

- Revised the logic of the `onContainerResize` function based on comments and discussions with @lukemelia . It now calculates the length of panels with a minimum length first before calculating the length of panels without a minimum length. This adjustment ensures a more proper distribution of the container size.
- Updated the IDs of the panels to start from 0 instead of 1 for consistency with array indices.
- Modified the `onResizeHandlerMouseMove` function to prevent the total length of the panels from exceeding the length of the container. I've made improvements involving the minimum length: if a user reduces the length of a panel, causing it to be lower than the minimum length, the panel will collapse. Conversely, if a user increases the length of a panel but it remains lower than the minimum length, the panel will open with the length set to the minimum length.


https://github.com/cardstack/boxel/assets/12637010/dcbb998e-d638-45bc-a8bb-c90c237af6f3

lukemelia commented on packages/boxel-ui/addon/src/components/resizable-panel-group/index.gts:
```diff
@@ -243,42 +242,78 @@ export default class ResizablePanelGroup extends Component<Signature> {
+    let secondElContext = this.listPanelContext.get(Number(secondElId));
+    if (!firstElContext || !secondElContext) {
+      console.warn(
+        'Expected firstElContext && nextElCosecondElContextntext to be defined',
```
typo: "nextElCosecondElContextntext" should be "secondElContext"

lukemelia commented on packages/boxel-ui/addon/src/components/resizable-panel-group/index.gts:
```diff
@@ -330,23 +370,24 @@ export default class ResizablePanelGroup extends Component<Signature> {
       );
     } else if (
-      buttonId.includes(String(this.listPanelContext.size - 1)) &&
+      buttonId.includes(String(this.listPanelContext.size - 2)) &&
```
I'm surprised to see a `-2` here. Is this right?

FadhlanR replied:
That's right because we start the panel id from 0 and the total number of resize handlers is consistently one less than the total number of panels, so the id of the last resize handler is equal to the total number of panels minus 2.

backspace replied:
I think these could benefit from extraction as named conditionals, like what is it describing? `isFirstButton` and `isLastButton`, for instance. It makes it easier to understand the various cases at high level.

FadhlanR committed:
Fix typo

backspace commented:
I’m confused by things in the Percy diff, like why would the Atom layout button not be in `main`?

![screencast 2023-11-22 09-38-30](https://github.com/cardstack/boxel/assets/43280/9f20a04e-b99f-4c24-a0f7-aa78fd97c7cf)

But I’ll look at and exercise the code now.

backspace commented on packages/boxel-ui/addon/src/components/resizable-panel-group/index.gts:
```diff
@@ -124,6 +116,29 @@ export default class ResizablePanelGroup extends Component<Signature> {
+        `Could not find selector: ${resizeHandlerSelector} when handling window resize for resizeable panel group`,
+      );
+      return undefined;
+    }
```
I know this is how it’s been for a while but maybe something like [`ember-ref-bucket`](https://github.com/lifeart/ember-ref-bucket) would be a useful modifier to track and pass elements around instead of going through the DOM.

FadhlanR replied:
Updated [1032fda](https://github.com/cardstack/boxel/pull/821/commits/1032fda4cc233dac3c0db7c2d9ae89bbd462ada3)

backspace commented on packages/boxel-ui/addon/src/components/resizable-panel-group/index.gts:
```diff
@@ -365,109 +448,77 @@ export default class ResizablePanelGroup extends Component<Signature> {
+        panelLengths[panelContext.id] = actualSize;
+      });
+    };
+    calculateLengtsOfPanelWithoutMinLegth();
```
```suggestion
    calculateLengthsOfPanelWithoutMinLength();
```

backspace reviewed (approved):
I left some refactoring suggestions but this looks good overall, and feels good too! 🎉

FadhlanR committed:
Update function name

Co-authored-by: Buck Doyle <buck.doyle@cardstack.com>

FadhlanR committed:
Update variables and functions name

FadhlanR committed:
Use ember-ref-bucket
